cmd/controller/app/controller.go

```diff
@@ -63,89 +66,145 @@ const controllerAgentName = "cert-manager"
+			if err != nil {
+				log.Error(err, "error creating leader election client")
+				return err
 			}
 
-			// don't run clusterissuers controller if scoped to a single namespace
-			if ctx.Namespace != "" && n == clusterissuers.ControllerName {
-				log.V(logf.InfoLevel).Info("not starting controller as cert-manager has been scoped to a single namespace")
-				continue
+			errorCh := make(chan error, 1)
+			if err := startLeaderElection(rootCtx, opts, leaderElectionClient, ctx.Recorder, leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(_ context.Context) {
+					close(elected)
+				},
+				OnStoppedLeading: func() {
+					select {
+					case <-rootCtx.Done():
+						// context was canceled, just return
+						return
+					default:
+						log.V(logf.ErrorLevel).Info("leader election lost")
+						errorCh <- errors.New("leader election lost")
+					}
+				},
+			}); err != nil {
+				return err
 			}
 
-			wg.Add(1)
-			iface, err := fn(ctx)
-			if err != nil {
-				log.Error(err, "error starting controller")
-				os.Exit(1)
+			select {
+			case err := <-errorCh:
+				return err
+			default:
+				return nil
 			}
-			go func(n string, fn controller.Interface) {
-				defer wg.Done()
-				log.V(logf.InfoLevel).Info("starting controller")
-
-				workers := 5
-				err := fn.Run(workers, stopCh)
-
-				if err != nil {
-					log.Error(err, "error starting controller")
-					os.Exit(1)
-				}
-			}(n, iface)
-		}
+		})
+	} else {
+		close(elected)
+	}
 
-		log.V(logf.DebugLevel).Info("starting shared informer factories")
-		ctx.SharedInformerFactory.Start(stopCh)
-		ctx.KubeSharedInformerFactory.Start(stopCh)
-		ctx.GWShared.Start(stopCh)
-		wg.Wait()
-		log.V(logf.InfoLevel).Info("control loops exited")
-		ctx.Metrics.Shutdown(metricsServer)
-		os.Exit(0)
+	select {
+	case <-rootCtx.Done(): // Exit early if the Elected channel gets closed because we are shutting down.
+		// Wait for error group to complete and return
+		return g.Wait()
+	case <-elected: // Don't launch the controllers unless we have been elected leader
+		// Continue with setting up controller
 	}
 
-	if !opts.LeaderElect {
-		run(context.TODO())
-		return
+	for n, fn := range controller.Known() {
+		log := log.WithValues("controller", n)
+
+		// only run a controller if it's been enabled
+		if !enabledControllers.Has(n) {
+			log.V(logf.InfoLevel).Info("not starting controller as it's disabled")
+			continue
+		}
+
+		// don't run clusterissuers controller if scoped to a single namespace
+		if ctx.Namespace != "" && n == clusterissuers.ControllerName {
+			log.V(logf.InfoLevel).Info("not starting controller as cert-manager has been scoped to a single namespace")
+			continue
+		}
+
+		iface, err := fn(ctx)
+		if err != nil {
+			log.Error(err, "error starting controller")
+
+			cancelContext()
+			_ = g.Wait() // Don't process errors, we already have an error
+			return err
```

Comment:
I hate throwing away errors. What errors would `g.Wait()` return here? Should we at least log them? Or add them to a `NewAggregate`

https://github.com/kubernetes/apimachinery/blob/a644435e2c133d990b624858d9c01985d7f59adf/pkg/util/errors/errors.go#L41-L61

Reply:
Note that g.Wait will also only return the first error that it encountered.

Reply:
I will aggregate these 2 errors.